src/forwarder.rs

```diff
@@ -0,0 +1,387 @@
+use astarte_device_sdk::{Aggregation, AstarteDeviceDataEvent};
+use edgehog_forwarder::astarte::{retrieve_connection_info, SessionInfo};
+use edgehog_forwarder::connections_manager::ConnectionsManager;
+use log::error;
+use reqwest::Url;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::task::JoinHandle;
+
+const CHANNEL_STATE_SIZE: usize = 50;
+
+#[derive(Debug)]
+struct Key(SessionInfo);
+
+impl Deref for Key {
+    type Target = SessionInfo;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Borrow<SessionInfo> for Key {
+    fn borrow(&self) -> &SessionInfo {
+        &self.0
+    }
+}
+
+impl PartialEq for Key {
+    fn eq(&self, other: &Self) -> bool {
+        self.host == other.host && self.port == other.port
+    }
+}
+
+impl Eq for Key {}
+
+impl Hash for Key {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.host.hash(state);
+        self.port.hash(state);
+        self.session_token.hash(state);
+    }
+}
+
+#[derive(Debug, Clone)]
+enum SessionStatus {
+    Connecting,
+    Connected,
+    Disconnected,
+}
+
+impl Display for SessionStatus {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Connecting => write!(f, "connecting"),
+            Self::Connected => write!(f, "connected"),
+            Self::Disconnected => write!(f, "disconnected"),
+        }
+    }
+}
+
+struct SessionState {
+    token: String,
+    status: SessionStatus,
+}
+
+/// Struct representing the state of a remote session with a device
+impl SessionState {
+    fn connecting(token: String) -> Self {
+        Self {
+            token,
+            status: SessionStatus::Connecting,
+        }
+    }
+
+    fn connected(token: String) -> Self {
+        Self {
+            token,
+            status: SessionStatus::Connected,
+        }
+    }
+
+    fn disconnected(token: String) -> Self {
+        Self {
+            token,
+            status: SessionStatus::Disconnected,
+        }
+    }
+}
+
+impl From<SessionState> for AstarteType {
+    fn from(value: SessionState) -> Self {
+        Self::String(value.status.to_string())
+    }
+}
+
+/// Device forwarder.
+///
+/// It maintains a collection of tokio task handles, each one identified by a [`Key`] containing
+/// the connection information and responsible for providing forwarder functionalities. For
+/// instance, a task could open a remote terminal between the device and a certain host.
+#[derive(Debug)]
+pub struct Forwarder {
+    tx_state: Sender<SessionState>,
+    tasks: HashMap<Key, JoinHandle<()>>,
+}
+
+impl Forwarder {
+    /// Initialize the forwarder instance, spawning also a task responsible for sending a property
+    /// necessary to update the device session state.
+    pub fn init<P>(publisher: P) -> Self
+    where
+        P: Publisher + 'static,
+    {
+        let (tx_state, rx_state) = channel::<SessionState>(CHANNEL_STATE_SIZE);
```

Comment:
Perhaps again `50` is too much, you are considering that at the same time 50 forwarders are trying to update their status?

Reply:
Yes maybe this number is a little bit high, since it is improbable to have 50 sessions opened at the same time and requesting a change of state. What do you think it could be a good estimate of the number of sessions on a given device? (Imo 5 could be enough)

Reply:
Yes I think 5 is reasonable number for starting.